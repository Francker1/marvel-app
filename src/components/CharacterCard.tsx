import { Link } from 'react-router-dom';
import styled from 'styled-components';
import HeartIconFilled from '../assets/Heart-icon-filled.svg';

interface CharacterCardProps {
  character: any;
  isFavorite: boolean;
}

const Card = styled.div`
  background-color: #fff;
  overflow: hidden;
  text-align: center;
  cursor: pointer;
  position: relative;
`;

const CharacterLink = styled(Link)`
  display: block;
  position: relative;

  &:after {
    content: '';
    display: block;
    width: 100%;
    height: 7px;
    background-color: red;
    position: absolute;
    bottom: 0;
    left: 0;
  }
`;

const CharacterImage = styled.img`
  width: 100%;
  height: 190px;
`;

const CharacterInfo = styled.div`
  height: 56px;
  background: linear-gradient(to bottom, red 0%, red 0%, transparent 0%);
  background-color: black;
  color: white;
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 16px;
  position: relative;
  transition: background 0.3s ease;

  ${Card}:hover & {
    background: linear-gradient(to bottom, red 0%, red 100%);
  }

  &:after {
    content: '';
    position: absolute;
    bottom: 0;
    right: 0;
    width: 0;
    height: 0;
    border-left: 12px solid transparent;
    border-bottom: 12px solid white;
  }
`;

const CharacterName = styled.span`
  font-size: 14px;
  flex: 1;
  text-align: left;
  text-transform: uppercase;
`;

const FavoritesIcon = styled.span`
  cursor: pointer;
  width: 14px;
  height: 14px;
  display: flex;
  align-items: center;
  justify-content: center;

  ${Card}:hover & {
    color: white;
  }

  img {
    width: 14px;
    height: 14px;
  }
`;

const CharacterCard: React.FC<CharacterCardProps> = ({ character, isFavorite }) => (
  <Card>
    <CharacterLink to={`/character/${character.id}`}>
      <CharacterImage src={`${character.thumbnail.path}.${character.thumbnail.extension}`} alt={character.name} />
    </CharacterLink>
    <CharacterInfo>
      <CharacterName>{character.name}</CharacterName>
      <FavoritesIcon onClick={(e) => {
        e.stopPropagation();
      }}>
        <img src={HeartIconFilled} alt="Add to Favorites" style={{ filter: isFavorite ? 'invert(1)' : 'none' }} />
      </FavoritesIcon>
    </CharacterInfo>
  </Card>
);

export default CharacterCard;
