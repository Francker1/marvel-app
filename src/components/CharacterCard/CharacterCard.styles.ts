import { Link } from 'react-router-dom';
import styled from 'styled-components';

export const Card = styled.div`
  background-color: #ffffff;
  overflow: hidden;
  text-align: center;
  cursor: pointer;
  position: relative;
`;

export const CharacterLink = styled(Link)`
  display: block;
  position: relative;

  &:after {
    content: '';
    display: block;
    width: 100%;
    height: 7px;
    background-color: #ec1d24;
    position: absolute;
    bottom: 0;
    left: 0;
  }
`;

export const CharacterImage = styled.img`
  width: 100%;
  height: 190px;
`;

export const CharacterInfo = styled.div`
  height: 56px;
  background-color: #000000;
  color: #ffffff;
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 16px;
  position: relative;
  transition: background 0.3s ease;

  ${Card}:hover & {
    background: #ec1d24;
  }

  &:after {
    content: '';
    position: absolute;
    bottom: 0;
    right: 0;
    width: 0;
    height: 0;
    border-left: 12px solid transparent;
    border-bottom: 12px solid #ffffff;
  }
`;

export const CharacterName = styled.span`
  font-size: 14px;
  flex: 1;
  text-align: left;
  text-transform: uppercase;
`;

export const FavoritesIcon = styled.span.attrs<{ isHovered: boolean }>((props) => ({
  isHovered: undefined,
}))<{ isHovered: boolean }>`
  cursor: pointer;
  width: 14px;
  height: 14px;
  display: flex;
  align-items: center;
  justify-content: center;

  ${Card}:hover & {
    color: white;
  }

  img {
    width: 14px;
    height: 14px;
  }
`;
