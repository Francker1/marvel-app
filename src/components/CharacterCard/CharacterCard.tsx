import { useState } from 'react';
import {
  Card,
  CharacterLink,
  CharacterImage,
  CharacterInfo,
  CharacterName,
  FavoritesIcon,
} from './CharacterCard.styles';
import HeartIconFilled from '../../assets/img/Heart-icon-filled.svg';
import HeartIconFilledWhite from '../../assets/img/Heart-icon-white.svg';
import HeartIconEmpty from '../../assets/img/Heart-icon-empty.svg';

interface CharacterCardProps {
  character: any;
  isFavorite: boolean;
  addFavorite: (character: any) => void;
  removeFavorite: (characterId: number) => void;
}

const CharacterCard: React.FC<CharacterCardProps> = ({
  character,
  isFavorite,
  addFavorite,
  removeFavorite,
}) => {
  const [isHovered, setIsHovered] = useState(false);

  const handleFavoriteClick = (e: React.MouseEvent) => {
    e.stopPropagation();
    if (isFavorite) {
      removeFavorite(character.id);
    } else {
      addFavorite(character);
    }
  };

  return (
    <Card
      onMouseEnter={() => setIsHovered(true)}
      onMouseLeave={() => setIsHovered(false)}
    >
      <CharacterLink to={`/character/${character.id}`}>
        <CharacterImage
          src={`${character.thumbnail.path}.${character.thumbnail.extension}`}
          alt={character.name}
        />
      </CharacterLink>
      <CharacterInfo>
        <CharacterName>{character.name}</CharacterName>
        <FavoritesIcon isHovered={isHovered} onClick={handleFavoriteClick}>
          <img
            src={
              isFavorite
                ? isHovered
                  ? HeartIconFilledWhite
                  : HeartIconFilled
                : HeartIconEmpty
            }
            alt="Add to Favorites"
          />
        </FavoritesIcon>
      </CharacterInfo>
    </Card>
  );
};

export default CharacterCard;
