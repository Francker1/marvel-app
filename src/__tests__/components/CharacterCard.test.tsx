import { render, screen, fireEvent } from '@testing-library/react';
import CharacterCard from '../../components/CharacterCard/CharacterCard';
import { BrowserRouter as Router } from 'react-router-dom';
import { expect, test, vi } from 'vitest';

const mockCharacter = {
  id: 1,
  name: 'Spider-Man',
  description: 'Friendly neighborhood Spider-Man',
  thumbnail: {
    path: 'http://example.com/spider-man',
    extension: 'jpg',
  },
};

test('renders character card correctly', () => {
  const { asFragment } = render(
    <Router>
      <CharacterCard
        character={mockCharacter}
        isFavorite={false}
        addFavorite={() => {}}
        removeFavorite={() => {}}
      />
    </Router>
  );
  expect(screen.getByText('Spider-Man')).toBeInTheDocument();
  expect(asFragment()).toMatchSnapshot(); // Crear snapshot
});

test('calls addFavorite on heart icon click when not favorite', () => {
  const addFavorite = vi.fn();
  render(
    <Router>
      <CharacterCard
        character={mockCharacter}
        isFavorite={false}
        addFavorite={addFavorite}
        removeFavorite={() => {}}
      />
    </Router>
  );
  const heartIcon = screen.getByAltText('Add to Favorites');
  fireEvent.click(heartIcon);
  expect(addFavorite).toHaveBeenCalledWith(mockCharacter);
});

test('calls removeFavorite on heart icon click when favorite', () => {
  const removeFavorite = vi.fn();
  render(
    <Router>
      <CharacterCard
        character={mockCharacter}
        isFavorite={true}
        addFavorite={() => {}}
        removeFavorite={removeFavorite}
      />
    </Router>
  );
  const heartIcon = screen.getByAltText('Add to Favorites');
  fireEvent.click(heartIcon);
  expect(removeFavorite).toHaveBeenCalledWith(mockCharacter.id);
});
