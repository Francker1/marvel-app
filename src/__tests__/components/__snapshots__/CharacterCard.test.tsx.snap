// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders character card correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-braxZu hxZAuk"
  >
    <a
      class="sc-gJhJTp hgPeRr"
      href="/character/1"
    >
      <img
        alt="Spider-Man"
        class="sc-elDIKY cMlytH"
        src="http://example.com/spider-man.jpg"
      />
    </a>
    <div
      class="sc-fQpRED hEHhrj"
    >
      <span
        class="sc-dsLQwm lnMvxE"
      >
        Spider-Man
      </span>
      <span
        class="sc-iKTcqh dSEPIF"
      >
        <img
          alt="Add to Favorites"
          src="/src/assets/img/Heart-icon-empty.svg"
        />
      </span>
    </div>
  </div>
</DocumentFragment>
`;
